linux/GPU.c

```diff
@@ -0,0 +1,272 @@
+in the source distribution for its full text.
+*/
+
+#include "linux/GPU.h"
+
+#include <assert.h>
+#include <ctype.h>
+#include <dirent.h>
+#include <errno.h>
+#include <sys/types.h>
+
+#include "XUtils.h"
+
+#include "linux/LinuxMachine.h"
+
+
+#define INVALID_CLIENT_ID ((unsigned long long int)-1)
+
+
+struct client_id {
+   char* pdev;
+   unsigned long long int id;
+   struct client_id* next;
+};
+
+enum section_state {
+   SECST_UNKNOWN,
+   SECST_DUPLICATE,
+   SECST_NEW,
+};
+
+static bool null_str_eq(const char* s1, const char* s2) {
+   if (s1 == NULL && s2 == NULL)
+      return true;
+
+   if (s1 && s2)
+      return String_eq(s1, s2);
+
+   return false;
+}
+
+static bool is_duplicate_client(const struct client_id* parsed, unsigned long long int id, const char* pdev) {
+   for (; parsed; parsed = parsed->next) {
+      if (id == parsed->id && null_str_eq(pdev, parsed->pdev)) {
+         return true;
+      }
+   }
+
+   return false;
+}
+
+static void update_system(LinuxProcessList* lpl, unsigned long long int time, char* engine) {
+   Machine* host = lpl->super.host;
+   LinuxMachine* lhost = (LinuxMachine*) host;
+   GPUEngineData** engineData = &lhost->gpuEngineData;
+
+   while (*engineData) {
+      if (String_eq((*engineData)->key, engine))
+         break;
+
+      engineData = &((*engineData)->next);
+   }
+
+   if (!*engineData) {
+      GPUEngineData* newData;
+
+      newData = xMalloc(sizeof(*newData));
+      *newData = (GPUEngineData) {
+         .prevTime = 0,
+         .curTime  = 0,
+         .key      = engine,
+         .next     = NULL,
+      };
+
+      *engineData = newData;
+   } else
+      free(engine);
+
+   (*engineData)->curTime += time;
+   lhost->curGpuTime += time;
+}
+
+/*
+ * Documentation reference:
+ * https://www.kernel.org/doc/html/latest/gpu/drm-usage-stats.html
+ */
+void GPU_readProcessData(LinuxProcessList* lpl, LinuxProcess* lp, openat_arg_t procFd) {
+   int fdinfoFd = -1;
+   DIR* fdinfoDir = NULL;
+   struct client_id* parsed_ids = NULL;
+   unsigned long long int new_gpu_time = 0;
+
+   fdinfoFd = Compat_openat(procFd, "fdinfo", O_RDONLY | O_NOFOLLOW | O_DIRECTORY | O_CLOEXEC);
+   if (fdinfoFd == -1)
+     goto out;
+
+   fdinfoDir = fdopendir(fdinfoFd);
+   if (!fdinfoDir)
+      goto out;
+   fdinfoFd = -1;
+
+#ifndef HAVE_OPENAT
+   char fdinfoPathBuf[32];
+   xSnprintf(fdinfoPathBuf, sizeof(fdinfoPathBuf), "/proc/%u/fdinfo", lp->super.pid);
+#endif
+
+   while (true) {
+      const struct dirent* entry;
+      const char* ename;
+      FILE* fp;
+      char* pdev = NULL;
+      unsigned long long int client_id = INVALID_CLIENT_ID;
+      enum section_state sstate = SECST_UNKNOWN;
+      char buf[256];
+
+      entry = readdir(fdinfoDir);
+      if (!entry)
+         break;
+      ename = entry->d_name;
+
+      if (String_eq(ename, ".") ||
+          String_eq(ename, "..") ||
+          String_eq(ename, "0") ||
+          String_eq(ename, "1") ||
+          String_eq(ename, "2"))
```

Comment:
Why exclude `stdin`, `stdout` and `stderr` explicitly?

Reply:
Yeah - had the same thought, I think this little optimization is not always valid unfortunately (e.g. a program closes one of these three at some point, as happens in some daemons, and then they later get reused).

Reply:
This reuse of standard file descriptors is dangerous, as any used library could write messages to stderr, corrupting any file opened at fileno 2. Do we really need to take care of such programs, which are mostly security related ones?

Reply:
Any library assuming it can write to stderr is dangerous.  :)  Its more common for libraries to use return codes to indicate failure (along with a strerror-style interface for custom errors).

But, maybe we're lucky and perhaps it is indeed the case that there's little overlap between the set of programs that close standard file descriptors and the set of programs that use GPUs (I expect the latter is mostly games or other interactive visual programs and AI/ML programs).

Reply:
It's bad practice for any library to assume any fd it didn't open is available. In particular writing to stdout and stderr is frowned upon.

Also including these fd is a matter of correctnes, while excluding the doesn't provide enough advantage to warrant the downsides.

Reply:
| [...] doesn't provide enough advantage to warrant the downsides.

Hmm - I expect there's a noticeable performance hit from opening/reading/closing these three files for every running process on a system, on every sample, for ~zero value.  I certainly understand why @cgzones went this path and maybe we should see if we can get away with it, since I suspect its unlikely to cause issues in practice.

Reply:
What about a different approach entirely:

Instead of skipping individual files, just mark in the process object if there were no files at all with any references for the DRM open. If so, skip such processes for a certain time (like ~5-10 seconds) from further scans. That way you reduce the amount of work to usually about 3-5% (quick local test gave 13 of ~390 processes). That's much more savings right away, with the only down-side that first GPU activity might be detected with a short delay (if not seen on first scan of a process). To contrast this: When skipping FDs 0, 1 and 2 you end up not scanning 5-10% (~380 FDs of 6440). Or written differently: That's doing still about 90% of work, where-as skipping processes without DRM access has about 5% of the work left.